src/consensus/consensus.md

```diff
@@ -0,0 +1,293 @@
+majority of the network. It is expected that **all nodes** that participate in the network
+are able to receive, validate and distribute transactions.
+
+The Mempool is the abstract component that stores transactions which are valid on top of
+the latest known Ledger State. Being valid implies in particular that the transaction is
+pending, i.e. it has not been included in any prior block ever, as in Cardano every valid
+transaction must consume at least one UTxO.
+
+Usually pending transactions are discarded when the node is shut down, however this is not
+a hard requirement. It is conceivable for a node to store the received pending
+transactions on the disk and re-evaluate them on startup. As IOG's Haskell implementation
+does not store the pending transactions on the disk, the mempool is currently considered
+part of the Consensus layer, and not part of the Storage layer, as it does not store data
+on the disk.
+
+The `TxSubmission` miniprotocol is the one used to diffuse the pending transactions
+through the network. Its state machine specification and general mechanics are described
+in the [Networking
+spec](https://ouroboros-network.cardano.intersectmbo.org/pdfs/network-spec/network-spec.pdf).
+
+Any mempool implementation has some requirements to allow diffusing transactions via
+`TxSubmission` and to fulfill the requirements of the Consensus layer in terms of forging
+blocks:
+- Acquiring a snapshot of valid transactions for a newly forged block should be as fast as
+  possible, as it will delay all other steps in the diffusion of such a block.
+- Index-based access to pending transactions, to fulfill the `TxSubmission` protocol requirements.
+- Re-validation on a new selected chain: it is useless to keep transactions that are no
+  longer valid in the mempool.
+
+## Forging new blocks
+
+Forging new blocks is the process of packing data that has not yet been included in the
+blockchain (pending transactions) into a fresh block. Blocks are signed by stake pool
+credentials, so only the nodes which are configured with such credentials strictly need to
+implement the forging mechanism.
+
+### The leadership check
+
+Every slot, the stake pool (whose keys the node was configured with) might be entitled to
+produce a block, depending on the rules of the particular Protocol that is running at the
+tip of the Cardano chain. For Praos, slots have a duration of one second. The probability
+of a node being elected in a slot is controlled via the _active slot coefficient_
+parameter (commonly referred to as `f`) which on mainnet is 0.05. This means that a block
+is expected once every 20 slots on average.
+
+In Ouroboros, a block consists of a sequence of transactions, so forging a block would
+look something similar to the following sequence of actions:
+
+1. Determine wheter my stake pool is entitled to create a block on this slot.
+2. Acquire a list of valid transactions with respect to the best Ledger State my node knows.
```

Comment:
Perhaps worth mentioning transactions should be selected w.r.t limits of block size, execution units